### PR TITLE
Use compiler intrinsics for `rz_swap_*` functions

### LINKED ISSUE #3286: Use `__builtin_bswapXX()` when available
To speed up produced code.

See the example on [GodBolt Compiler Explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYMQAZgCspBwAyeAyYAHLuAEaYxCAA7F6kAA6oCoS2DM5uHt5%2Byak2AkEh4SxRMfEWmFYFDEIETMQEme6evpXV6XUNBEVhkdFxCQr1jc3ZbcPdvSVlgwCUFqiuxMjsHACkAExewchuWADU616Ow/ioAHQIx9jrGgCCWzsMe66Hx6cE%2BIJXN3ePD0mNmQB2CtGCmAOrmCBC4ADZjAQDsQAF7GBQAdyYiWMrlhcIg0ME8MRBwAbmI5kdYgAhf53ACcxEwBGWDAOEAptCOJw%2BBwAHFT1rFHByuTzbl5sAK5sc6Q9hQARemA%2BrA0EMcEhDX40mo9FYnHDeEQGEkpFcoW0lUM8XHRUczliHmOPmCo6bOEHDSqABivo0Gn9gatoqgdqlNxlHq9PuDQd9sq88oBTJZbLFzo%2BfPhocz3L%2BUoOubl9NiyoVqqY6rBEKhMK8mz1aMx2NxBEbhIbTYtlOpKcZEYd4azvJO0a2sb9AYTIepYadBcjRfdk%2B908D/qTA40adZxHZi5dObheaPhelJeTZYrAPuQLwINr2u7zYNbeGndfvdoVp3ttHYdzzHUVV09dd43jM9gMlaUwKnSDE1LSs9wzYDXXHXN50dCNYOLU8Ywg/0kOvBVyxVe81UfDUtUhIkCDhSQ31bHE8UYrtBEY0lLX7G0hxw0cMNAoVwLjGdEM3Gc5ACAJoNwqN4PXSTlNnBNpNk5DU34kclyE/CRIQ4jA0k%2BNA3UuTRzwrC1x9YyE1M%2BzfXMzTGWZfdDwjPTGzPeSi28lzyMrSjq2o59IRhLikX1Fj0QYyRTU4pifz/PjAIE3S3QMoiVIkjRzOwnSJQUrLbNU3L/Wc0itLSwrs0wgibOnBziJnMyZIspcrIa0TVDslq%2BuIyr/zctDPL5byCpgqMJsalqWqGm8KPo80oRNFKHgOTb62JBEkSYHkHR9SRJH5PwDgiA711O47NK25F0wPZEW0NdsTSYKkAFonvfVj8QgCJt0WoL6MbUlXE7db7jukGewOfb7XXY7ToZHxNmkc7Lp9NGUeuyRbq2kbHuil68U7d6Di%2B4m21JzZ/sBsjb3%2BejIqhdjIehiKkrhzHVCRnwfAZNG4Thfnjs2TZSAxhGsbRyR%2BeF3H8c2wn2Sp372PJynnupuK6YCxmHmW3bztcNbeI2rajdJeGvEO3mTrOi7pdURWqrulWDlMCJoVoGwGGMCIWLez7PYDn2/YDoOCQB/Wlu/E2IfNqHLfjm27b5lG0clp3bfXbGfFdlN3Ye9kvfD4JI8NMmQ7LugI8DqvaZjqqlTjxLSW9tmk459u9p5vmBaFkWfDFiWpdzmXjvluFC/%2BYv3ND7264rhvsQ1muw%2BX/3V8Sdjm4HQK7xhA4WCYYIEr24hgHMA5kAQBoACoH7hq%2ByXZlOduty7TYJenk82mGHdv4QyVttOKpIQQI1cGzUBiRiAwioBALYmw07ri2D4VQ2ceboMwbfS66DaC0FUOsHwjgGDIMlkwbOktkB/x7rqPaXBLrezWqAwBSIIhMIRt7EBbsP7gKRMgLhudO7xToVtOBCCkHiyYMIu2ODs5yLQZsDBNClEEKISQshFC4ZcEUWo8RysS7en1hwBYtBOA%2BF4J4DgWhSCoE4KKBQSwViQieDwUgBBNBmIWAAa28F4C4XgGQMn5MLBkjZ%2BQaC4BoWI%2BhOCSGsd4%2BxnBeAKBABoTx3iFhwFgEgNALBEh0GiOQSgBSin0BiMgYAXBxakCwGSR8mAABqeBMAYgAPKJEYJwDxNBfbRHSf9ZJERggNAAJ69N4KM5gxBxkdIiNoTA1gpmkAKWwQQHTNSTNsbwLAp8jDiF2fUvAzJrB4DJJgdJxzMCqGWXiNYHiYRVGSeCCIxAJnOCwMkgg8CWBTIWFQAwwAFCtPaV0np3BeD8EECIMQ7ApAyEEIoFQ6hjm6D0QYIwIBTDGHMG89JkAFioESDUa5H0ADqYhaBpKqMsmo9gGBOBcC0EAcJ/BMumP0GIXBYjSDyGkAQYxPDsoFTULlpQBi8ukJYelnQRhNBZdkdlsrzkCC6I0CVsxpUWAVcKtluqpgQhmFKvlCxnHLFWBIcxliknHIcRwA4Ls4QfUYgca%2BIJakXE2ByXAhASAei8FwOYvAvG7LmH4gJ3rhaxBCfELwYSfBwg0NICxHBEmkBsXYh1aSMlZPDaQXJiAQBLAIIkPEpSIDlOKcQUIrA1jOtdZId1yBPWbAuB4zA%2BAiDwPQHoGFwhRDiERQOlFahkkYtIBiD5iQAXxI4FYzNySHUdLxOWpEqAqBOrCU2ltbbvUcmcIUmtgbg2huyZGrwgSr03tvXe%2BdGas28BzRYPNYatARvnZsO12bUn5o/QsS5xBUh2EkEAA):

See comparison of two functions:
```c
uint64_t bu64() {
    uint64_t a = 0x4485592466554422, b = 0x224455668544;
    return __builtin_bswap64(a) - __builtin_bswap64(b);
}
```
vs
```c
uint64_t u64() {
    uint64_t a = 0x4485592466554422, b = 0x224455668544;
    return rz_swap_ut64(a) - rz_swap_ut64(b);
}
```

### With `rz_swap()` function:
```asm
       push    rbp
        mov     rbp, rsp
        push    rbx
        sub     rsp, 16
        movabs  rax, 4937450579352175650
        mov     QWORD PTR [rbp-16], rax
        movabs  rax, 37676885902660
        mov     QWORD PTR [rbp-24], rax
        mov     rax, QWORD PTR [rbp-16]
        mov     rdi, rax
        call    rz_swap_ut64
        mov     rbx, rax
        mov     rax, QWORD PTR [rbp-24]
        mov     rdi, rax
        call    rz_swap_ut64
        mov     rdx, rbx
        sub     rdx, rax
        mov     rax, rdx
        mov     rbx, QWORD PTR [rbp-8]
        leave
        ret
```
### With `__builtin_bswap()` function:
```asm
        push    rbp
        mov     rbp, rsp
        movabs  rax, 4937450579352175650
        mov     QWORD PTR [rbp-8], rax
        movabs  rax, 37676885902660
        mov     QWORD PTR [rbp-16], rax
        mov     rax, QWORD PTR [rbp-8]
        bswap   rax
        mov     rdx, rax
        mov     rax, QWORD PTR [rbp-16]
        bswap   rax
        sub     rdx, rax
        mov     rax, rdx
        pop     rbp
        ret
```
### With `rz_swap()` function:
```asm
        stp     x29, x30, [sp, -48]!
        mov     x29, sp
        str     x19, [sp, 16]
        mov     x0, 17442
        movk    x0, 0x6655, lsl 16
        movk    x0, 0x5924, lsl 32
        movk    x0, 0x4485, lsl 48
        str     x0, [sp, 40]
        mov     x0, 34116
        movk    x0, 0x5566, lsl 16
        movk    x0, 0x2244, lsl 32
        str     x0, [sp, 32]
        ldr     x0, [sp, 40]
        bl      rz_swap_ut64
        mov     x19, x0
        ldr     x0, [sp, 32]
        bl      rz_swap_ut64
        sub     x0, x19, x0
        ldr     x19, [sp, 16]
        ldp     x29, x30, [sp], 48
        ret
```
### With `__builtin_bswap()` function:
```asm
        sub     sp, sp, #16
        mov     x0, 17442
        movk    x0, 0x6655, lsl 16
        movk    x0, 0x5924, lsl 32
        movk    x0, 0x4485, lsl 48
        str     x0, [sp, 8]
        mov     x0, 34116
        movk    x0, 0x5566, lsl 16
        movk    x0, 0x2244, lsl 32
        str     x0, [sp]
        ldr     x0, [sp, 8]
        rev     x1, x0
        ldr     x0, [sp]
        rev     x0, x0
        sub     x0, x1, x0
        add     sp, sp, 16
        ret
```

### With `rz_swap()` function:
```asm
        sub     sp, sp, #48
        stp     x29, x30, [sp, #32]             // 16-byte Folded Spill
        add     x29, sp, #32
        mov     x8, #17442
        movk    x8, #26197, lsl #16
        movk    x8, #22820, lsl #32
        movk    x8, #17541, lsl #48
        stur    x8, [x29, #-8]
        mov     x8, #34116
        movk    x8, #21862, lsl #16
        movk    x8, #8772, lsl #32
        str     x8, [sp, #16]
        ldur    x0, [x29, #-8]
        bl      rz_swap_ut64
        str     x0, [sp, #8]                    // 8-byte Folded Spill
        ldr     x0, [sp, #16]
        bl      rz_swap_ut64
        mov     x8, x0
        ldr     x0, [sp, #8]                    // 8-byte Folded Reload
        subs    x0, x0, x8
        ldp     x29, x30, [sp, #32]             // 16-byte Folded Reload
        add     sp, sp, #48
        ret
```

### With `__builtin_bswap()` function:
```asm
        sub     sp, sp, #16
        mov     x8, #17442
        movk    x8, #26197, lsl #16
        movk    x8, #22820, lsl #32
        movk    x8, #17541, lsl #48
        str     x8, [sp, #8]
        mov     x8, #34116
        movk    x8, #21862, lsl #16
        movk    x8, #8772, lsl #32
        str     x8, [sp]
        ldr     x8, [sp, #8]
        rev     x8, x8
        ldr     x9, [sp]
        rev     x9, x9
        subs    x0, x8, x9
        add     sp, sp, #16
        ret
```

### With `rz_swap()` function:
```asm
        push    {r4, r5, r7, r8, r9, lr}
        sub     sp, sp, #16
        add     r7, sp, #0
        adr     r3, .L15
        ldrd    r2, [r3]
        strd    r2, [r7, #8]
        adr     r3, .L15+8
        ldrd    r2, [r3]
        strd    r2, [r7]
        ldrd    r0, [r7, #8]
        bl      rz_swap_ut64
        mov     r4, r0
        mov     r5, r1
        ldrd    r0, [r7]
        bl      rz_swap_ut64
        mov     r2, r0
        mov     r3, r1
        subs    r8, r4, r2
        sbc     r9, r5, r3
        mov     r2, r8
        mov     r3, r9
        mov     r0, r2
        mov     r1, r3
        adds    r7, r7, #16
        mov     sp, r7
        pop     {r4, r5, r7, r8, r9, pc}
```

### With `__builtin_bswap()` function:
```asm
        push    {r4, r5, r6, r7}
        sub     sp, sp, #16
        add     r7, sp, #0
        adr     r3, .L23
        ldrd    r2, [r3]
        strd    r2, [r7, #8]
        adr     r3, .L23+8
        ldrd    r2, [r3]
        strd    r2, [r7]
        ldr     r3, [r7, #8]
        rev     r2, r3
        ldr     r3, [r7, #12]
        rev     r3, r3
        mov     r0, r3
        mov     r1, r2
        ldr     r3, [r7]
        rev     r6, r3
        ldr     r3, [r7, #4]
        rev     ip, r3
        mov     r2, ip
        mov     r3, r6
        subs    r4, r0, r2
        sbc     r5, r1, r3
        mov     r2, r4
        mov     r3, r5
        mov     r0, r2
        mov     r1, r3
        adds    r7, r7, #16
        mov     sp, r7
        pop     {r4, r5, r6, r7}
        bx      lr
```

### With `rz_swap()` function:
```asm
       .quad   .L.u64,.TOC.@tocbase,0
.L.u64:
        mflr 0
        std 0,16(1)
        std 30,-16(1)
        std 31,-8(1)
        stdu 1,-144(1)
        mr 31,1
        addis 9,2,.LC0@toc@ha
        ld 9,.LC0@toc@l(9)
        std 9,112(31)
        addis 9,2,.LC1@toc@ha
        ld 9,.LC1@toc@l(9)
        std 9,120(31)
        ld 3,112(31)
        bl rz_swap_ut64
        mr 30,3
        ld 3,120(31)
        bl rz_swap_ut64
        mr 9,3
        subf 9,9,30
        mr 3,9
        addi 1,31,144
        ld 0,16(1)
        mtlr 0
        ld 30,-16(1)
        ld 31,-8(1)
        blr
        .long 0
        .byte 0,0,0,1,128,2,0,1
```

### With `__builtin_bswap()` function:
```asm
        .quad   .L.bu64,.TOC.@tocbase,0
.L.bu64:
        std 31,-8(1)
        stdu 1,-80(1)
        mr 31,1
        addis 9,2,.LC0@toc@ha
        ld 9,.LC0@toc@l(9)
        std 9,48(31)
        addis 9,2,.LC1@toc@ha
        ld 9,.LC1@toc@l(9)
        std 9,56(31)
        addi 10,31,48
        li 9,4
        lwbrx 8,9,10
        lwbrx 10,0,10
        sldi 8,8,32
        or 10,10,8
        addi 9,31,56
        li 8,4
        lwbrx 7,8,9
        lwbrx 9,0,9
        sldi 7,7,32
        or 9,9,7
        subf 9,9,10
        mr 3,9
        addi 1,31,80
        ld 31,-8(1)
        blr
        .long 0
        .byte 0,0,0,0,128,1,0,1
```

### With `rz_swap()` function:
```asm
        addiu   $sp,$sp,-56
        sw      $31,52($sp)
        sw      $fp,48($sp)
        sw      $17,44($sp)
        sw      $16,40($sp)
        move    $fp,$sp
        li      $2,1716846592                 # 0x66550000
        ori     $2,$2,0x4422
        li      $3,1149566976                 # 0x44850000
        ori     $3,$3,0x5924
        sw      $2,24($fp)
        sw      $3,28($fp)
        li      $2,1432748032                 # 0x55660000
        ori     $2,$2,0x8544
        li      $3,8772           # 0x2244
        sw      $2,32($fp)
        sw      $3,36($fp)
        lw      $4,24($fp)
        lw      $5,28($fp)
        jal     rz_swap_ut64
        nop

        move    $16,$2
        move    $17,$3
        lw      $4,32($fp)
        lw      $5,36($fp)
        jal     rz_swap_ut64
        nop

        move    $4,$2
        move    $5,$3
        subu    $2,$16,$4
        sltu    $6,$16,$2
        subu    $3,$17,$5
        subu    $4,$3,$6
        move    $3,$4
        move    $sp,$fp
        lw      $31,52($sp)
        lw      $fp,48($sp)
        lw      $17,44($sp)
        lw      $16,40($sp)
        addiu   $sp,$sp,56
        jr      $31
        nop
```
### With `__builting_bswap()` function:
```asm
        addiu   $sp,$sp,-56
        sw      $31,52($sp)
        sw      $fp,48($sp)
        sw      $17,44($sp)
        sw      $16,40($sp)
        move    $fp,$sp
        li      $2,1716846592                 # 0x66550000
        ori     $2,$2,0x4422
        li      $3,1149566976                 # 0x44850000
        ori     $3,$3,0x5924
        sw      $2,24($fp)
        sw      $3,28($fp)
        li      $2,1432748032                 # 0x55660000
        ori     $2,$2,0x8544
        li      $3,8772           # 0x2244
        sw      $2,32($fp)
        sw      $3,36($fp)
        lw      $4,24($fp)
        lw      $5,28($fp)
        jal     __bswapdi2
        nop

        move    $16,$2
        move    $17,$3
        lw      $4,32($fp)
        lw      $5,36($fp)
        jal     __bswapdi2
        nop

        move    $4,$2
        move    $5,$3
        subu    $2,$16,$4
        sltu    $6,$16,$2
        subu    $3,$17,$5
        subu    $4,$3,$6
        move    $3,$4
        move    $sp,$fp
        lw      $31,52($sp)
        lw      $fp,48($sp)
        lw      $17,44($sp)
        lw      $16,40($sp)
        addiu   $sp,$sp,56
        jr      $31
        nop
```

As you could see from the assembly, it should speedup loading and analyzing big files in situations where endianness of the analyzed file is different from the endianness (or bitness) of the host machine.

See relevant functions (`rz_swap*()`, `rz_read*()`, `rz_write()`) in `librz/include/rz_endian.h`

Also move `test_endian()` function from `test/unit/test_util.c` to `test/unit/test_endian.c`, and add testing for 64 bit numbers, as well as add tests for byte swapping functions.

### PR DESCRIPTION
**Pull request by rizan21**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Use `__builtin_bswap`, when available, for swapping bytes in `rz_swap_*` API.

**Test plan**

Add tests for `rz_swap_*` API in `unit/test_endian.c`.

**Closing issues**

Fixes #3286.

### CODE REVIEW

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good, thank you! As you might notice, apart from those `rz_swap_*()` functions there are more in the same file that could use those builtins as well, e.g. some of the `rz_read_*()` or `rz_write_*()` functions.

**Comment by wargio** [comment]
also i think this could be a good opportunity to add the check directly into meson.

**Comment by rizan21** [comment]
I'm not sure what the intrinsic functions for `rz_read_*()` and `rz_write()_*` are. What should I use for them?

**Comment by XVilka** [comment]
@rizan21 could you please use the `cc.has_function('__builtin_bswap16')` Meson function for detection instead? And put it here: https://github.com/rizinorg/rizin/blob/dev/meson.build

**Comment by rizan21** [comment]
Hey @XVilka and @wargio  , I've added the below snippet to `meson.build`
```meson
foreach intrin : [
      ['__builtin_bswap16', 'BSWAP16'],
      ['__builtin_bswap32', 'BSWAP32'],
      ['__builtin_bswap64', 'BSWAP64'],
    ]
    has_intrin = it_cc.has_function(intrin[0])
    define_name = 'HAVE_' + intrin[1]
    it_userconf.set10(define_name, has_intrin)
endforeach
```
and was expecting `HAVE_BSWAPXX` to be available in `rz_endian.h`. so that the below code would work
```c++
#if HAVE_BSWAPXX
#define rz_swap_utXX __builtin_bswapXX
#else
static inline utXX rz_swap_utXX(utXX val) {
    // relevant code
}
#endif
```
But it doesn't seem to be working and the intrinsic functions are never being used. I've gone through a lot of `meson` syntax and can't seem to figure it out. Would appreciate some help :)

**Comment by XVilka** [comment]
@rizan21 it's because this code write this file `librz/include/rz_userconf.h.in` which will become `librz/include/rz_userconf.h`. Thus, you need to include it in the `rz_endian.h`, I suppose. @ret2libc am I right?

File: librz/include/rz_endian.h [hunk 1]
```diff
@@ -591,37 +592,68 @@ static inline void rz_write_ble_double(void *src, double val, bool big_endian) {
 	big_endian ? rz_write_be_double(src, val) : rz_write_le_double(src, val);
 }
 
+#ifndef __has_builtin
```
Comment (XVilka) [inline librz/include/rz_endian.h]: You forgot to remove this

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/include/rz_endian.h (hunk 1 above)
Comment (rizan21) [reply librz/include/rz_endian.h]: I'm afk for a few hours, I'll get it fixed asap.

Comment (rizan21) [reply librz/include/rz_endian.h]: Done

File: meson.build [hunk 2]
```diff
@@ -467,6 +467,16 @@ foreach it : ccs
     it_userconf.set10('HAVE_HEADER_@0@'.format(hdr.underscorify().to_upper()), ok)
   endforeach
 
+  foreach intrin : [
```
Comment (ret2libc) [inline meson.build]: There is already a foreach loop that does this so you can add these functions there

Comment (rizan21) [reply meson.build]: Should I rename the macro to have the same name as the function (followed by all other macros) or do I keep them the same as before ?.  ( i.e. `HAVE_BSWAP16` -> `HAVE___BUILTIN_BSWAP16` ?)

Comment (XVilka) [reply meson.build]: Yeah, `HAVE___BUILTIN_BSWAP16` is better

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
Thank you!

**Comment by rizan21** [comment]
Was a pleasure contributing :)

**Pull request merged**